src/builder/QuarkBuilder.sol

```diff
@@ -353,52 +522,38 @@ contract QuarkBuilder {
-        Actions.Action[] memory transferActions,
-        Accounts.ChainAccounts[] memory chainAccountsList,
-        uint256 targetChainId,
-        uint256 plannedBridgeAmountToTargetChain
-    ) internal pure {
-        uint256 paymentTokensUsed = 0;
-        for (uint256 i = 0; i < transferActions.length; ++i) {
-            Actions.TransferActionContext memory transferActionContext =
-                abi.decode(transferActions[i].actionContext, (Actions.TransferActionContext));
-            // Filter with the targetChainId and paymentTokensUsed will track on one chain at a time
-            if (transferActionContext.chainId == targetChainId) {
-                address transferToken = transferActionContext.token;
-                uint256 transferAmount = transferActionContext.amount;
-                uint256 paymentAssetBalanceOnChain = Accounts.sumBalances(
-                    Accounts.findAssetPositions(transferToken, transferActions[i].chainId, chainAccountsList)
-                );
-                paymentTokensUsed += transferActions[i].paymentMaxCost;
-                if (transferToken == transferActions[i].paymentToken) {
-                    // If the payment token is the transfer token and this is the target chain, we need to account for the transfer amount
-                    // If its transfer step, check if user has enough balance to cover the transfer amount after bridge
-                    if (
-                        paymentAssetBalanceOnChain + plannedBridgeAmountToTargetChain
-                            < paymentTokensUsed + transferAmount
-                    ) {
-                        revert MaxCostTooHigh();
-                    }
+        for (uint256 i = 0; i < nonBridgeActions.length; ++i) {
+            Actions.Action memory nonBridgeAction = nonBridgeActions[i];
+            if (nonBridgeAction.chainId != targetChainId) {
+                revert InvalidActionChain();
+            }
+            paymentTokenCost += nonBridgeAction.paymentMaxCost;
 
-                    // Special handling as the payment token is sent out so it will be part of the cost
-                    paymentTokensUsed += transferAmount;
-                } else {
-                    // Just check payment token can cover the max cost
-                    if (paymentAssetBalanceOnChain < paymentTokensUsed) {
-                        revert MaxCostTooHigh();
-                    }
+            if (Strings.stringEqIgnoreCase(nonBridgeAction.actionType, Actions.ACTION_TYPE_TRANSFER)) {
+                Actions.TransferActionContext memory transferActionContext =
+                    abi.decode(nonBridgeAction.actionContext, (Actions.TransferActionContext));
+                if (transferActionContext.token == paymentToken) {
+                    paymentTokenCost += transferActionContext.amount;
                 }
+            } else if (Strings.stringEqIgnoreCase(nonBridgeAction.actionType, Actions.ACTION_TYPE_SUPPLY)) {
```

Comment:
nit: we are generating all these action types so why would it ever be lowercase and require ignoring case?

Reply:
sure, it's probably unnecessary